File: src/lib/network.ts
import { authed } from "@/lib/auth";

/**
 * Client for the research hub.
 *
 * Every call goes through `authed`, which carries the access token and renews
 * it once on a 401 — the directory and the connection graph are members-only,
 * so there is no anonymous path here to fall back to.
 */

export type Role = "student" | "mentor";
export type ConnectionStatus = "pending" | "accepted" | "declined";

/**
 * Where the viewer stands with a person.
 *
 * Computed on the server and returned on the card, so a list of sixty people
 * renders sixty correct buttons without the client joining two arrays.
 */
export type Standing =
  | "none"
  | "self"
  | "connected"
  | "awaiting_them"
  | "awaiting_you"
  | "declined";

export interface PersonCard {
  id: number;
  handle: string;
  display_name: string;
  role: Role;
  institution: string | null;
  headline: string | null;
  interests: string[];
  open_to_mentoring: boolean;
  is_demo: boolean;
  joined_at: string;
  standing: Standing;
  connection_id: number | null;
  /** Only filled for people you are connected to. */
  modules_completed: number | null;
  badges_earned: number | null;
}

export interface DirectoryResponse {
  people: PersonCard[];
  total: number;
  query: string | null;
  role: Role | null;
}

export interface ConnectionRecord {
  id: number;
  status: ConnectionStatus;
  note: string | null;
  created_at: string;
  responded_at: string | null;
  /** True when you are the one who asked. */
  outgoing: boolean;
  person: PersonCard;
}

export interface NetworkResponse {
  connections: ConnectionRecord[];
  incoming: ConnectionRecord[];
  outgoing: ConnectionRecord[];
  pending_incoming: number;
}

export function fetchDirectory(options: {
  q?: string;
  role?: Role | null;
  mentorsOnly?: boolean;
} = {}) {
  const params = new URLSearchParams();
  if (options.q?.trim()) params.set("q", options.q.trim());
  if (options.role) params.set("role", options.role);
  if (options.mentorsOnly) params.set("mentors_only", "true");
  const query = params.toString();
  return authed<DirectoryResponse>(`/api/network/directory${query ? `?${query}` : ""}`);
}

export function fetchNetwork() {
  return authed<NetworkResponse>("/api/network");
}

export function sendRequest(handle: string, note?: string) {
  return authed<ConnectionRecord>("/api/network/requests", "POST", {
    handle,
    note: note?.trim() || null,
  });
}

export function acceptRequest(id: number) {
  return authed<ConnectionRecord>(`/api/network/requests/${id}/accept`, "POST", {});
}

export function declineRequest(id: number) {
  return authed<ConnectionRecord>(`/api/network/requests/${id}/decline`, "POST", {});
}

export function withdrawRequest(id: number) {
  return authed<void>(`/api/network/requests/${id}`, "DELETE");
}

export function removeConnection(id: number) {
  return authed<void>(`/api/network/connections/${id}`, "DELETE");
}

export function fetchMyCard() {
  return authed<PersonCard>("/api/network/me");
}

export function updateMyCard(body: {
  role?: Role;
  headline?: string | null;
  interests?: string | null;
  open_to_mentoring?: boolean;
}) {
  return authed<PersonCard>("/api/network/me", "PATCH", body);
}

File: backend/app/models/network.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

"""
Wire shapes for the research hub.

Everything here is deliberately narrower than the database row. A directory
listing is public to signed-in members, so it carries a handle and a headline
and never an email address — a hub where browsing people hands out their inbox
is a spam list, not a directory.
"""

Role = Literal["student", "mentor"]
Status = Literal["pending", "accepted", "declined"]
"""How the viewer stands with a person, from the viewer's side of the row."""
Standing = Literal["none", "self", "connected", "awaiting_them", "awaiting_you", "declined"]


def _tidy(value: str | None) -> str | None:
    if value is None:
        return None
    collapsed = " ".join(value.split())
    return collapsed or None


class PersonCard(BaseModel):
    """One row of the directory."""

    id: int
    handle: str
    display_name: str
    role: Role
    institution: str | None = None
    headline: str | None = None
    interests: list[str] = Field(default_factory=list)
    open_to_mentoring: bool = False
    # The expanded profile, shown when a row is opened. Optional throughout —
    # a card with none of it set still renders, just shorter.
    position: str | None = None
    education: str | None = None
    focus: str | None = None
    mentoring: str | None = None
    availability: str | None = None
    # A seeded example profile so a fresh install has a directory to browse.
    # Surfaced so the UI can say so on the card — a fixture that looks like a
    # real academic is the one thing this feature must not ship.
    is_demo: bool = False
    # Where the card's picture lives, if it has one. The client falls back to
    # initials, so absent is an ordinary state rather than a broken one.
    avatar_url: str | None = None
    # Seniority in the directory, and the only timestamp a stranger sees.
    joined_at: datetime
    # How the person asking stands with this one, so the card can render its
    # own button without the client cross-referencing two lists.
    standing: Standing = "none"
    # The connection row behind a non-"none" standing, for accept/withdraw.
    connection_id: int | None = None
    # Only ever filled for people the viewer is connected to. Progress is
    # private until you are.
    modules_completed: int | None = None
    badges_earned: int | None = None


class DirectoryResponse(BaseModel):
    people: list[PersonCard]
    total: int
    # Echoed back so the client can tell a stale response from a current one.
    query: str | None = None
    role: Role | None = None


class ConnectRequest(BaseModel):
    model_config = ConfigDict(extra="ignore")

    handle: str = Field(min_length=1, max_length=32)
    note: str | None = Field(default=None, max_length=500)

    @field_validator("note")
    @classmethod
    def _clean_note(cls, value: str | None) -> str | None:
        return _tidy(value)


class ConnectionRecord(BaseModel):
    id: int
    status: Status
    note: str | None = None
    created_at: datetime
    responded_at: datetime | None = None
    # True when the viewer is the one who asked.
    outgoing: bool
    # The other party — never the viewer, so a client never has to work out
    # which of two ids is "them".
    person: PersonCard


class NetworkResponse(BaseModel):
    """Everything the hub's own page needs, in one round trip."""

    connections: list[ConnectionRecord]
    incoming: list[ConnectionRecord]
    outgoing: list[ConnectionRecord]
    # Shown on the nav badge.
    pending_incoming: int


class ProfileCard(BaseModel):
    """The viewer's own directory card, as they can edit it."""

    model_config = ConfigDict(extra="ignore")

    role: Role | None = None
    headline: str | None = Field(default=None, max_length=140)
    interests: str | None = Field(default=None, max_length=240)
    open_to_mentoring: bool | None = None

    @field_validator("headline", "interests")
    @classmethod
    def _clean(cls, value: str | None) -> str | None:
        return _tidy(value)
